icu4c/source/common/umutablecptrie.cpp

```diff
@@ -1543,7 +1543,7 @@ int32_t MutableCodePointTrie::compactTrie(int32_t fastILimit, UErrorCode &errorC
     MixedBlocks mixedBlocks;
     int32_t newDataLength = compactData(fastILimit, newData, newDataCapacity,
                                         dataNullIndex, mixedBlocks, errorCode);
-    if (U_FAILURE(errorCode)) { return 0; }
+    if (U_FAILURE(errorCode)) { uprv_free(newData); return 0; }
```

Comment:
same as comment above, fix this by changing 
```
    uint32_t *newData = (uint32_t *)uprv_malloc(newDataCapacity * 4);
```
to
```
    LocalMemory<uint32_t> newData;
    newData.allocateInsteadAndCopy(newDataCapacity);
```
see similar code in https://source.chromium.org/chromium/chromium/src/+/main:third_party/icu/source/tools/genrb/parse.cpp;l=980?q=allocateInsteadAndCopy&ss=chromium%2Fchromium%2Fsrc

Reply:
@FrankYFTang like this?:
```
--- a/icu4c/source/common/umutablecptrie.cpp
+++ b/icu4c/source/common/umutablecptrie.cpp
@@ -1531,7 +1531,8 @@ int32_t MutableCodePointTrie::compactTrie(int32_t fastILimit, UErrorCode &errorC
         errorCode = U_MEMORY_ALLOCATION_ERROR;
         return 0;
     }
-    uint32_t* newData = static_cast<uint32_t*>(uprv_malloc(newDataCapacity * 4));
+    LocalMemory<uint32_t> buffer;
+    uint32_t *newData = buffer.allocateInsteadAndCopy(newDataCapacity * 4);
     if (newData == nullptr) {
         errorCode = U_MEMORY_ALLOCATION_ERROR;
         return 0;
```

Reply:
No, that doesn't work, this gives me an error when running `make check` ☹.